File: lib/core/services/request_log.dart
import 'dart:collection';

import 'package:flutter_riverpod/flutter_riverpod.dart';

/// One backend round-trip as seen by the app: enough to debug "why does the
/// home screen show —" without reading device logs.
class RequestLogEntry {
  RequestLogEntry({
    required this.startedAt,
    required this.method,
    required this.path,
    required this.query,
    this.statusCode,
    this.durationMs,
    this.error,
    this.summary,
    this.bytes,
  });

  final DateTime startedAt;
  final String method;
  final String path;
  final Map<String, dynamic> query;
  final int? statusCode;
  final int? durationMs;
  final String? error;

  /// Free-form one-liner: selected source, run, counts.
  final String? summary;
  final int? bytes;

  bool get ok => error == null && (statusCode ?? 0) < 400;

  String get queryString => query.entries
      .map((e) => '${e.key}=${Uri.encodeQueryComponent('${e.value}')}')
      .join('&');
}

/// Ring buffer of the most recent backend requests.
class RequestLog extends StateNotifier<UnmodifiableListView<RequestLogEntry>> {
  RequestLog() : super(UnmodifiableListView(const []));

  static const _max = 60;
  final _entries = <RequestLogEntry>[];

  /// Global hook so the network layer (no Riverpod access) can record.
  static RequestLog? _active;
  static RequestLog? get active => _active;

  static bool enabled = true;

  void attach() => _active = this;

  void add(RequestLogEntry e) {
    if (!enabled) return;
    _entries.insert(0, e);
    if (_entries.length > _max) _entries.removeRange(_max, _entries.length);
    state = UnmodifiableListView(List.of(_entries));
  }

  void clear() {
    _entries.clear();
    state = UnmodifiableListView(const []);
  }

  static void record(RequestLogEntry e) => _active?.add(e);
}

final requestLogProvider =
    StateNotifierProvider<RequestLog, UnmodifiableListView<RequestLogEntry>>(
  (ref) {
    final log = RequestLog();
    log.attach();
    return log;
  },
);

File: src/core/services/requestLog.ts
/// One backend round-trip as seen by the app — port of
/// `lib/core/services/request_log.dart`. Enough to debug "why does the home
/// screen show —" without reading device logs. Ring buffer of the most recent
/// backend requests, backed by a zustand store.

import { create } from "zustand";

export interface RequestLogEntry {
  id: number;
  startedAt: Date;
  method: string;
  path: string;
  query: Record<string, unknown>;
  statusCode?: number | null;
  durationMs?: number | null;
  error?: string | null;
  /// Free-form one-liner: selected source, run, counts.
  summary?: string | null;
}

export function requestLogOk(entry: RequestLogEntry): boolean {
  return !entry.error && (entry.statusCode ?? 0) < 400;
}

export function requestQueryString(query: Record<string, unknown>): string {
  return Object.entries(query)
    .map(([key, value]) => `${key}=${encodeURIComponent(String(value))}`)
    .join("&");
}

const MAX_ENTRIES = 60;

interface RequestLogState {
  entries: RequestLogEntry[];
  enabled: boolean;
  record: (entry: Omit<RequestLogEntry, "id">) => void;
  setEnabled: (enabled: boolean) => void;
  clear: () => void;
}

let nextId = 1;

export const useRequestLogStore = create<RequestLogState>((set) => ({
  entries: [],
  enabled: true,
  record: (entry) =>
    set((state) => {
      if (!state.enabled) return state;
      const withId: RequestLogEntry = { ...entry, id: nextId++ };
      return { entries: [withId, ...state.entries].slice(0, MAX_ENTRIES) };
    }),
  setEnabled: (enabled) => set({ enabled }),
  clear: () => set({ entries: [] }),
}));

export function recordRequest(entry: Omit<RequestLogEntry, "id">): void {
  useRequestLogStore.getState().record(entry);
}
